### PR TITLE
Always cache the full result

### LINKED ISSUE #4178: DDC-2043: Extra cache operation in DBAL\Cache\ResultCacheStatement.php
Jira issue originally created by user bogdan.albei:

This is the closeCursor() method in DBAL\Cache\ResultCacheStatement.php:

```
public function closeCursor()
    {
        $this->statement->closeCursor();
        if ($this->emptied && $this->data !== null) {
            $data = $this->resultCache->fetch($this->cacheKey);
            if ( ! $data) {
                $data = array();
            }
            $data[$this->realKey] = $this->data;

            $this->resultCache->save($this->cacheKey, $data, $this->lifetime);
            unset($this->data);
        }
    }
```

We are using Memcache and I noticed an extra GET operation on all cache misses. In the code above I believe the fetch call is not necessary and that the code would do the same without it. 
Also, may I ask why is the SQL used as a key in the cached data?

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | bugfix, improvement
| BC Break     | no

Fixes https://github.com/doctrine/dbal/issues/4178. Closes https://github.com/doctrine/dbal/pull/4189.

1. Cache the whole result right after it's available regardless of how it's fetched and whether it's fetched at all. The original implementation of caching only after the result is fully fetched optimizes for the edge case and shouldn't exist.
2. Serve the result from an array result which is the same as if the result was retrieved from the cache.
3. Move the result caching logic from `CachingResult` to the wrapper connection. This way, we can remove the redundant call to the cache.
4. Remove `CachingResult` as no longer used.
5. Remove `testDontFinishNoCache()` as no longer valid.
6. Rename `ResultCacheTest` since it covers the whole feature, not a specific class.

### CODE REVIEW

File: src/Connection.php [hunk 1]
```diff
@@ -1070,23 +1069,32 @@ public function executeCacheQuery($sql, $params, $types, QueryCacheProfile $qcp)
...
         if ($item->isHit()) {
-            $data = $item->get();
+            $value = $item->get();
             // is the real key part of this row pointers map or is the cache only pointing to other cache keys?
-            if (isset($data[$realKey])) {
-                $result = new ArrayResult($data[$realKey]);
-            } elseif (array_key_exists($realKey, $data)) {
+            if (isset($value[$realKey])) {
+                $result = new ArrayResult($value[$realKey]);
+            } elseif (array_key_exists($realKey, $value)) {
```
Comment (morozov) [inline src/Connection.php]: This line isn't covered by the tests and looks redundant. An empty result should be represented in the cache as an empty array, not as NULL. Even if some old caches contain NULL, a requirement to clear the cache during an upgrade doesn't look like a BC break.

It was introduced in https://github.com/doctrine/dbal/commit/dea79e7bfbabf973e807539e258ccf3b9ee98f45. Although the commit adds new tests, reverting this condition doesn't break any of the existing ones.

**Review by greg0ire — COMMENTED** [review]
I think `docs/en/reference/caching.rst` will need to be changed, since users no longer need to explicitly read the entire result set from the database. We should instead warn them that using the caching feature does that implicitly. Maybe we should also document it as a minor BC break in the upgrade guide?

**Review by beberlei — DISMISSED** [review]

**Comment by morozov** [comment]
> I think `docs/en/reference/caching.rst` will need to be changed, since users no longer need to explicitly read the entire result set from the database.

Thanks! Updated.

> We should instead warn them that using the caching feature does that implicitly.

Should we? IMO the current implementation is the most reasonable and doesn't need documenting technical details.

> Maybe we should also document it as a minor BC break in the upgrade guide?

What is the BC break?

**Review by greg0ire — APPROVED** [review]
> What is the BC break?

The result set is now implicitly read entirely. If somebody uses this API to read the first row of a query that returns many results, they might have memory issues. But As I write this lines, I realize this is too far-fetched :sweat_smile: . Ignore me.

**Pull request merged**